# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/indexedDB/ActorsParent.cpp
# Commit: dffe1e6c58bd
# Full Hash: dffe1e6c58bd717cda6184a647c54b0bc7aca12e
# Author: Simon Giesecke <sgiesecke@mozilla.com>
# Date: 2019-11-05 21:49:23
# Regressor Bug: 1168606
# File Overlap Count: 1
# Description:
#   Bug 1168606 - Extract functions for accessing sort key into Cursor. r=ytausky,asuth
#   
#   Depends on D44443
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D45308
# ==============================================================================

diff -r 41f1bef0f019 -r dffe1e6c58bd dom/indexedDB/ActorsParent.cpp
--- a/dom/indexedDB/ActorsParent.cpp	Tue Nov 05 12:47:28 2019 +0000
+++ b/dom/indexedDB/ActorsParent.cpp	Tue Nov 05 12:47:36 2019 +0000
@@ -7685,6 +7685,9 @@
 
   void SetOptionalKeyRange(const Maybe<SerializedKeyRange>& aOptionalKeyRange,
                            bool* aOpen);
+
+  const Key& GetSortKey() const;
+  Key* GetModifiableSortKey();
 };
 
 class Cursor::CursorOpBase : public TransactionDatabaseOperationBase {
@@ -15132,7 +15135,7 @@
     return false;
   }
 
-  const Key& sortKey = IsLocaleAware() ? mLocaleAwarePosition : mPosition;
+  const Key& sortKey = GetSortKey();
 
   switch (aParams.type()) {
     case CursorRequestParams::TContinueParams: {
@@ -15419,6 +15422,12 @@
   return IPC_OK();
 }
 
+const Key& Cursor::GetSortKey() const {
+  return IsLocaleAware() ? mLocaleAwarePosition : mPosition;
+}
+
+Key* Cursor::GetModifiableSortKey() { return &const_cast<Key&>(GetSortKey()); }
+
 /*******************************************************************************
  * FileManager
  ******************************************************************************/
@@ -25765,8 +25774,10 @@
 nsresult Cursor::CursorOpBase::PopulateResponseFromStatement(
     mozIStorageStatement* const aStmt, const bool aInitializeResponse) {
   Transaction()->AssertIsOnConnectionThread();
-  MOZ_ASSERT(aInitializeResponse ==
-             (mResponse.type() == CursorResponse::T__None));
+  MOZ_ASSERT_IF(aInitializeResponse,
+                mResponse.type() == CursorResponse::T__None);
+  MOZ_ASSERT_IF(!aInitializeResponse,
+                mResponse.type() != CursorResponse::T__None);
   MOZ_ASSERT_IF(
       mFiles.IsEmpty() &&
           (mResponse.type() ==
@@ -25913,10 +25924,7 @@
   // For unique cursors, we need to skip records with the same key. The SQL
   // queries currently do not filter these out.
   Key previousKey =
-      IsUnique(mCursor->mDirection)
-          ? (mCursor->IsLocaleAware() ? mCursor->mLocaleAwarePosition
-                                      : mCursor->mPosition)
-          : Key{};
+      IsUnique(mCursor->mDirection) ? mCursor->GetSortKey() : Key{};
 
   uint32_t extraCount = 0;
   do {
@@ -25938,9 +25946,7 @@
     }
 
     if (IsUnique(mCursor->mDirection)) {
-      const auto& currentKey = mCursor->IsLocaleAware()
-                                   ? mCursor->mLocaleAwarePosition
-                                   : mCursor->mPosition;
+      const auto& currentKey = mCursor->GetSortKey();
       const bool sameKey = previousKey == currentKey;
       previousKey = currentKey;
       if (sameKey) {
@@ -26190,7 +26196,7 @@
   // ContinueOp::DoDatabaseWork.
   //
   // TODO: We should somehow evaluate the effects of this. Maybe use a smaller
-  // extra count that for ContinueOp?
+  // extra count than for ContinueOp?
   //
   // TODO: If this is done here, do this in the other Do*DatabaseWork functions
   // as well (or move this to DoDatabaseWork).
@@ -26618,8 +26624,7 @@
   // key/position in the operation's result. Maybe rename to
   // targetKey/targetPosition (which is also not exact, as it might imply that
   // the result always has this key).
-  Key& currentKey = mCursor->IsLocaleAware() ? mCursor->mLocaleAwarePosition
-                                             : mCursor->mPosition;
+  Key& currentKey = *mCursor->GetModifiableSortKey();
 
   IDB_LOG_MARK_PARENT_TRANSACTION_REQUEST(
       "PRELOAD: ContinueOp: currentKey == %s", "currentKey",
